File: src/Java/myservicebus-postgresql/src/main/java/com/myservicebus/persistence/postgresql/PostgreSqlInboxStore.java
package com.myservicebus.persistence.postgresql;

import com.myservicebus.persistence.InboxAcquisition;
import com.myservicebus.persistence.InboxMessageKey;
import com.myservicebus.persistence.InboxStore;
import com.myservicebus.persistence.InboxTransaction;
import com.myservicebus.persistence.OutboxWriter;
import com.myservicebus.tasks.CancellationToken;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;

public final class PostgreSqlInboxStore implements InboxStore {
    private final Connection connection;
    private final PostgreSqlOutboxWriter outbox;

    public PostgreSqlInboxStore(Connection connection, String serviceName) {
        this.connection = Objects.requireNonNull(connection, "connection");
        this.outbox = new PostgreSqlOutboxWriter(connection, serviceName);
    }

    @Override
    public CompletableFuture<InboxTransaction> acquire(InboxMessageKey key, CancellationToken cancellationToken) {
        try {
            cancellationToken.throwIfCancelled();
            requireTransaction();
            String sql = """
                    INSERT INTO myservicebus.inbox_message (consumer_scope, message_id, state)
                    VALUES (?, ?, 0) ON CONFLICT (consumer_scope, message_id) DO NOTHING;
                    """;
            try (PreparedStatement statement = connection.prepareStatement(sql)) {
                statement.setString(1, key.consumerScope());
                statement.setObject(2, key.messageId());
                if (statement.executeUpdate() == 1) {
                    return CompletableFuture.completedFuture(
                            new Transaction(connection, outbox, key, InboxAcquisition.ACQUIRED));
                }
            }

            String stateSql = """
                    SELECT state FROM myservicebus.inbox_message
                    WHERE consumer_scope = ? AND message_id = ?;
                    """;
            try (PreparedStatement statement = connection.prepareStatement(stateSql)) {
                statement.setString(1, key.consumerScope());
                statement.setObject(2, key.messageId());
                try (ResultSet result = statement.executeQuery()) {
                    if (!result.next()) {
                        throw new IllegalStateException("The conflicting inbox record was not found.");
                    }
                    InboxAcquisition acquisition = result.getShort(1) == 1
                            ? InboxAcquisition.COMPLETED
                            : InboxAcquisition.IN_PROGRESS;
                    return CompletableFuture.completedFuture(new Transaction(connection, outbox, key, acquisition));
                }
            }
        } catch (Exception failure) {
            return CompletableFuture.failedFuture(failure);
        }
    }

    private void requireTransaction() throws SQLException {
        if (connection.isClosed() || connection.getAutoCommit()) {
            throw new IllegalStateException("An active caller-owned PostgreSQL transaction is required.");
        }
    }

    private static final class Transaction implements InboxTransaction {
        private final Connection connection;
        private final OutboxWriter outbox;
        private final InboxMessageKey key;
        private final InboxAcquisition acquisition;

        private Transaction(
                Connection connection,
                OutboxWriter outbox,
                InboxMessageKey key,
                InboxAcquisition acquisition) {
            this.connection = connection;
            this.outbox = outbox;
            this.key = key;
            this.acquisition = acquisition;
        }

        @Override
        public InboxMessageKey getKey() {
            return key;
        }

        @Override
        public InboxAcquisition getAcquisition() {
            return acquisition;
        }

        @Override
        public OutboxWriter getOutbox() {
            return outbox;
        }

        @Override
        public CompletableFuture<Void> complete() {
            if (acquisition != InboxAcquisition.ACQUIRED) {
                return CompletableFuture.failedFuture(
                        new IllegalStateException("Only an acquired inbox message can be completed."));
            }
            try {
                if (connection.isClosed() || connection.getAutoCommit()) {
                    return CompletableFuture.failedFuture(
                            new IllegalStateException("The caller-owned PostgreSQL transaction is no longer active."));
                }
            } catch (SQLException failure) {
                return CompletableFuture.failedFuture(failure);
            }
            String sql = """
                    UPDATE myservicebus.inbox_message
                    SET state = 1, completed_at_utc = CURRENT_TIMESTAMP
                    WHERE consumer_scope = ? AND message_id = ? AND state = 0;
                    """;
            try (PreparedStatement statement = connection.prepareStatement(sql)) {
                statement.setString(1, key.consumerScope());
                statement.setObject(2, key.messageId());
                if (statement.executeUpdate() != 1) {
                    throw new IllegalStateException("The acquired inbox message could not be completed.");
                }
                return CompletableFuture.completedFuture(null);
            } catch (Exception failure) {
                return CompletableFuture.failedFuture(failure);
            }
        }

        @Override
        public void close() {
        }
    }
}

File: src/MyServiceBus.PostgreSql/PostgreSqlInboxStore.cs
using MyServiceBus.Persistence;
using Npgsql;
using NpgsqlTypes;

namespace MyServiceBus.Persistence.PostgreSql;

public sealed class PostgreSqlInboxStore : IInboxStore
{
    private readonly NpgsqlConnection connection;
    private readonly NpgsqlTransaction transaction;
    private readonly PostgreSqlOutboxWriter outbox;

    public PostgreSqlInboxStore(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        string serviceName)
    {
        this.connection = connection ?? throw new ArgumentNullException(nameof(connection));
        this.transaction = transaction ?? throw new ArgumentNullException(nameof(transaction));
        if (!ReferenceEquals(transaction.Connection, connection))
            throw new ArgumentException("The transaction must belong to the supplied connection.", nameof(transaction));
        outbox = new PostgreSqlOutboxWriter(connection, transaction, serviceName);
    }

    public async Task<IInboxTransaction> AcquireAsync(
        InboxMessageKey key,
        CancellationToken cancellationToken = default)
    {
        if (transaction.Connection is null)
            throw new InvalidOperationException("The PostgreSQL transaction is no longer active.");

        const string sql = """
            INSERT INTO myservicebus.inbox_message (consumer_scope, message_id, state)
            VALUES (@consumer_scope, @message_id, 0)
            ON CONFLICT (consumer_scope, message_id) DO NOTHING;
            """;
        await using var command = new NpgsqlCommand(sql, connection, transaction);
        command.Parameters.AddWithValue("consumer_scope", NpgsqlDbType.Text, key.ConsumerScope);
        command.Parameters.AddWithValue("message_id", NpgsqlDbType.Uuid, key.MessageId);
        var inserted = await command.ExecuteNonQueryAsync(cancellationToken) == 1;
        if (inserted)
            return new InboxTransaction(connection, transaction, outbox, key, InboxAcquisition.Acquired);

        const string stateSql = """
            SELECT state
            FROM myservicebus.inbox_message
            WHERE consumer_scope = @consumer_scope AND message_id = @message_id;
            """;
        await using var stateCommand = new NpgsqlCommand(stateSql, connection, transaction);
        stateCommand.Parameters.AddWithValue("consumer_scope", NpgsqlDbType.Text, key.ConsumerScope);
        stateCommand.Parameters.AddWithValue("message_id", NpgsqlDbType.Uuid, key.MessageId);
        var stateValue = await stateCommand.ExecuteScalarAsync(cancellationToken);
        if (stateValue is null or DBNull)
            throw new InvalidOperationException("The conflicting inbox record was not found.");
        var state = Convert.ToInt16(stateValue);
        var acquisition = state == 1 ? InboxAcquisition.Completed : InboxAcquisition.InProgress;
        return new InboxTransaction(connection, transaction, outbox, key, acquisition);
    }

    private sealed class InboxTransaction : IInboxTransaction
    {
        private readonly NpgsqlConnection connection;
        private readonly NpgsqlTransaction transaction;

        public InboxTransaction(
            NpgsqlConnection connection,
            NpgsqlTransaction transaction,
            IOutboxWriter outbox,
            InboxMessageKey key,
            InboxAcquisition acquisition)
        {
            this.connection = connection;
            this.transaction = transaction;
            Outbox = outbox;
            Key = key;
            Acquisition = acquisition;
        }

        public InboxMessageKey Key { get; }
        public InboxAcquisition Acquisition { get; }
        public IOutboxWriter Outbox { get; }

        public async Task CompleteAsync(CancellationToken cancellationToken = default)
        {
            if (Acquisition != InboxAcquisition.Acquired)
                throw new InvalidOperationException("Only an acquired inbox message can be completed.");
            if (transaction.Connection is null)
                throw new InvalidOperationException("The PostgreSQL transaction is no longer active.");

            const string sql = """
                UPDATE myservicebus.inbox_message
                SET state = 1, completed_at_utc = CURRENT_TIMESTAMP
                WHERE consumer_scope = @consumer_scope AND message_id = @message_id AND state = 0;
                """;
            await using var command = new NpgsqlCommand(sql, connection, transaction);
            command.Parameters.AddWithValue("consumer_scope", NpgsqlDbType.Text, Key.ConsumerScope);
            command.Parameters.AddWithValue("message_id", NpgsqlDbType.Uuid, Key.MessageId);
            if (await command.ExecuteNonQueryAsync(cancellationToken) != 1)
                throw new InvalidOperationException("The acquired inbox message could not be completed.");
        }

        public ValueTask DisposeAsync() => ValueTask.CompletedTask;
    }
}
